cooperative_lanechange/src/cooperative_lanechange.cpp

```diff
@@ -0,0 +1,867 @@
+        
+        auto shortest_path2 = temp_route.get().shortestPath();
+
+        //To find downtrack- distance of veh2 from current to start of end lanelet + downtrack of veh1 in its own current lanelet
+        lanelet::BasicPoint2d p = shortest_path2.back().centerline2d().front(); //Start point of last lanelet in path
+        
+        double downtrack_1 = wm_->routeTrackPos(p).downtrack;
+        double remaining_downtrack = downtrack_1 - veh2_downtrack + ego_current_downtrack;
+        //double downtrack_1 = map_graph->routeTrackPos(p).downtrack;
+        current_gap = remaining_downtrack;
+
+
+        return current_gap;
+
+    }
+
+    void CooperativeLaneChangePlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+    }
+    void CooperativeLaneChangePlugin::twist_cd(const geometry_msgs::TwistStampedConstPtr& msg)
+    {
+        current_speed_ = msg->twist.linear.x;
+    }
+    
+    void CooperativeLaneChangePlugin::bsm_cb(const cav_msgs::BSMConstPtr& msg)
+    {
+        bsm_core_ = msg->core_data;
+        
+    }
+
+    void CooperativeLaneChangePlugin::run()
+    {
+    	initialize();
+        ros::CARMANodeHandle::setSpinRate(10.0);
+        ros::CARMANodeHandle::spin();
+
+    }
+
+    
+
+    bool CooperativeLaneChangePlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest &req, cav_srvs::PlanTrajectoryResponse &resp){
+        
+        //Check if Lane Change is needed
+        std::vector<cav_msgs::Maneuver> maneuver_plan;
+        for(const auto maneuver : req.maneuver_plan.maneuvers)
+        {
+            if(maneuver.type == cav_msgs::Maneuver::LANE_CHANGE)
+            {
+                maneuver_plan.push_back(maneuver);
+            }
+        }
+
+        //Current only checking for first lane change maneuver message
+        int target_lanelet_id = stoi(maneuver_plan[0].lane_change_maneuver.ending_lane_id);
+        double target_downtrack = maneuver_plan[0].lane_change_maneuver.end_dist;
+        //get subject vehicle info
+        lanelet::BasicPoint2d veh_pos(req.vehicle_state.X_pos_global, req.vehicle_state.Y_pos_global);
+        double current_downtrack = wm_->routeTrackPos(veh_pos).downtrack;
+        auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, veh_pos, 10);       
+        int current_lanelet_id = current_lanelets[0].second.id();
+        if(current_lanelet_id == target_lanelet_id && current_downtrack >= target_downtrack - destination_range_){
+            cav_msgs::LaneChangeStatus lc_status_msg;
+            lc_status_msg.status = cav_msgs::LaneChangeStatus::PLANNING_SUCCESS;
+            //No description as per UI documentation
+            lanechange_status_pub_.publish(lc_status_msg);
+        }
+
+        std::vector<cav_msgs::RoadwayObstacle> rwol = wm_->getRoadwayObjects();
+        //Assuming only one object in list 
```

Comment:
I'm not sure this is a valid assumption, we might have multiple objects, such as a vehicle in front of us, or a BSM from a vehicle behind us, being tracked. It might be worth thinking about how to identify that vehicle in the other lane.